src/Maestro/Maestro.Common/AzureDevOpsTokens/AzureDevOpsTokenProvider.cs

```diff
@@ -26,42 +28,123 @@ namespace Maestro.Common.AzureDevOpsTokens;
+        var account = m.Groups["account"].Value;
+        return await GetTokenForAccountAsync(account);
+    }
+
+    private TokenCredential GetCredential(string account)
+    {
+        if (_accountCredentials.TryGetValue(account, out var credential))
         {
-            return (await credential.GetTokenAsync(new TokenRequestContext([AzureDevOpsScope]))).Token;
+            return credential;
         }
 
-        // We can also define just one MI for all accounts
-        if (_tokenCredentials.TryGetValue("default", out var defaultCredential))
+        if (_accountCredentials.TryGetValue("default", out var defaultCredential))
         {
-            return (await defaultCredential.GetTokenAsync(new TokenRequestContext([AzureDevOpsScope]))).Token;
+            return defaultCredential;
         }
 
         throw new ArgumentOutOfRangeException(
             $"Azure DevOps account {account} does not have a configured PAT or credential. " +
             $"Please add the account to the 'AzureDevOps.Tokens' or 'AzureDevOps.ManagedIdentities' configuration section");
     }
+
+    private static Dictionary<string, TokenCredential> GetCredentials(
+        AzureDevOpsTokenProviderOptions options,
+        Func<string, TokenRequestContext, CancellationToken, string> patResolver)
+    {
+        Dictionary<string, TokenCredential> credentials = [];
+
+        foreach (var pair in options)
+        {
+            var account = pair.Key;
+            var option = pair.Value;
+
+            // 1. AzDO PAT from configuration
+            if (!string.IsNullOrEmpty(option.Token))
+            {
+                credentials[account] = new ResolvingCredential((context, cancellationToken) => patResolver(account, context, cancellationToken));
+                continue;
+            }
+
+            // 2. Federated token that can be used to fetch an app token (for CI scenarios)
+            if (!string.IsNullOrEmpty(option.FederatedToken))
+            {
+                credentials[account] = AppCredential.CreateFederatedCredential(option.AppId, option.FederatedToken!);
+                continue;
+            }
+
+            // 3. Managed identity (for server-to-AzDO scenarios)
+            if (!string.IsNullOrEmpty(option.ManagedIdentityId))
+            {
+                credentials[account] = option.ManagedIdentityId == "system"
+                    ? new ManagedIdentityCredential()
+                    : new ManagedIdentityCredential(option.ManagedIdentityId);
+                continue;
+            }
+
+            // 4. Azure CLI authentication setup by the caller (for CI scenarios)
+            if (option.DisableInteractiveAuth)
+            {
+                credentials[account] = AppCredential.CreateNonUserCredential(option.AppId);
+                continue;
+            }
+
+            // 5. Interactive login (user-based scenario)
+            credentials[account] = new DefaultAzureCredential(includeInteractiveCredentials: true);
```

Comment:
Does this also cache the auth request (i.e. the user doesn't get prompted for creds every time we need a token)?

Reply:
This does not seem to open any browser window at all and silently logs you in

Reply:
Not sure though

Reply:
> It might have caching depending on the credential that gets used. For example the environment variable credential can build e.g. a ClientSecretCredential, which uses an instance of MSAL ConfidentialClientApplication internally (see [source](https://github.com/Azure/azure-sdk-for-net/blob/b33a5e3096df23bd5a9af0210428f08ffe75a8eb/sdk/identity/Azure.Identity/src/ClientSecretCredential.cs#L18)). MSAL has an in-memory cache for tokens

Reply:
> This does not seem to open any browser window at all and silently logs you in

Yeah, since it uses the `VisualStudio*` credentials. If you log out of both Studio and Code it would then use interactive which in my experience results in getting prompted every time

Reply:
Doesn't it only use the VS one when you F5 the Darc project though? If I run it from my terminal which is what most do, it uses some other maybe?

Reply:
Regardless, I think it would be rather something for you to figure out when removing the AzDO token from local flows. But thisv PR rather aims to do that on the server side mainly with also laying some foundations for the other flows.

Reply:
Nope, it uses it for any invocation, so guessing the creds are stored somewhere and the class is just accessing them

Reply:
> Regardless, I think it would be rather something for you to figure out when removing the AzDO token from local flows

Yeah, absolutely, just wanted to ask if maybe you would know